scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +172,393 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
         i += 1
 
     return np.reshape(x, size1d)
+
+
+class FastNumericalInverse():
```

Comment:
I named it after the paper "Continuous random variate generation by _fast numerical inversion_". Would the other methods fit well in this interface (e.g. selected by a 'method' parameter)?

Reply:
I think the interface would be convoluted if we combine different methods: while there is some overlap of the parameters, each method has some unique parameters

Reply:
OK

Reply:
> I am not sure about the name. In UNU.RAN, there are three numerical inversion methods:
> 
>     * PINV (polynomial interpolation, usually fastest method)
> 
>     * HINV (Hermite interpolation, this method)
> 
>     * NINV (other numerical inversions based on bisection, Newton etc)

I agree with @chrisb83 here. It would be nice to rename this if we were to add other variants of approximation like inversion using Polynomial interpolation. Maybe, call it `FastHermiteNumericalInversion` so if we have a variant implementing the Polynomial interpolation based inversion, we could name it `FastPolynomialNumericalInversion`.

Other thing is that: Do we want more of these methods in Python? If we are integrating UNU.RAN, I guess its performance would mostly be better than Python + we won't have to maintain our own version of the method. In case we don't add newer methods like this in Python, the current name sounds fine.

Reply:
I was actually thinking that UNU.RAN would just replace the implementation of the algorithm, which only accounts for about 10% of this PR. The algorithm is really simple, as far as algorithms go. Most of the work of this PR went into documentation, user interface (e.g. getting RVS and QRVS broadcasting right), tests, and the benchmark. 

So I think this class should be named however you want to name the UNU.RAN version, and if the UNU.RAN version is significantly faster, you basically just replace the private function `_fast_numerical_inverse` with the UNU.RAN version. This would add the monotonicity check and make it easy to accept new arguments that allow the user to specify the domain and mesh points. If you want to avoid the overhead of scipy.stats distributions, you can make the first parameter more flexible - instead of requiring it to be a frozen distribution, you could accept any object with `cdf`, `pdf`, `ppf`, and `isf` methods, or you could accept a tuple with those methods, or whatever.

Reply:
> So I think this class should be named however you want to name the UNU.RAN version, and if the UNU.RAN version is significantly faster, you basically just replace the private function `_fast_numerical_inverse` with the UNU.RAN version. This would add the monotonicity check and make it easy to accept new arguments that allow the user to specify the domain and mesh points.

I haven't yet decided what to name the UNU.RAN versions. But it makes sense to replace this with UNU.RAN version to add more functionality to it. 



> If you want to avoid the overhead of scipy.stats distributions, you can make the first parameter more flexible - instead of requiring it to be a frozen distribution, you could accept any object with `cdf`, `pdf`, `ppf`, and `isf` methods, or you could accept a tuple with those methods, or whatever. (Or maybe with UNU.RAN you don't even need the user to provide some of these methods anymore.)

That's exactly what I was thinking. In the prototype, I accept whatever functions are needed as arguments. The idea about accepting a class also sounds nice. Thanks!